libevmasm/RuleList.h

```diff
@@ -374,6 +390,30 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart7(
 		false
 	});
 
+	rules.push_back({
+		// AND(A, SHR(B, X)) -> SHR(B, X) if 2^B-1 == A
+		{Instruction::AND, {A, {Instruction::SHR, {B, X}}}},
+		[=]() -> Pattern { return {Instruction::SHR, {B, X}}; },
+		false,
+		[=]() {
+			if (B.d() > 256)
+				return false;
+			return (u256(2) << (256 - static_cast<int>(B.d())-1)) - 1 == A.d();
+			}
+	});
+
+	rules.push_back({
+		// AND(SHR(A, X), B) -> SHR(A, X) if 2^B-1 == A
```

Comment:
Is that the same rule, just the arguments of `and` reversed? Do we need that?

Reply:
> Is that the same rule, just the arguments of and reversed? 

Yes

> Do we need that?

Depends. If the optimizer rearranges the arguments and runs several iterations then maybe not.

Reply:
Ah right, we used to have a step that would automatically match in reverse if the operation is commutative, but that was removed, so we have to explicitly add these rules now.

Reply:
Can we at least swap A and B in the rule pattern here and re-use the same lambda function as condition for both rules that way?

Reply:
Sure. I did that initially because I couldn't get it to compile and I thought maybe the order mattered ;)

Reply:
I'm not entirely sure whether the order matters myself, but I think it shouldn't...